### PR TITLE
Use integer division trick to get exact base64 length

### PR DESCRIPTION
**Pull request by gfoidl**
...and avoid (explicit) modulo operation.

Has better [codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzMMA2gFkYGABYQAJgElxkgBQ37T12MkB5MT4ILlwWAEEAc0jYXFxeBRhnLkleLjTIgEoAXVNzMSgE7AwYJnIkBjSMBgBxWwAhPBhUABkYLkj7dyrKrjEODDaO+0y8sxNqcymGAHoZgwgJGEcGADMoUQY7DAwxXBA5yN57DmAWSHwZxwgMLlsZqG4+ARngSQhgGYAOUjBsJC+5DAwGwYC+AE5HIDtNhtCptOCAOwwOhgJBIACszTAdAx5GWSERM1wUDAM1SwCg2EKqhmzBYAAVCgpioIRLAWrxPiSycwZiIuIkoBhzrgAMQtUgY7R0MbTd4dBgQAbSLgMAC8DHc7gVWTS/UG7U6dkyswYugAVAwUGpptM5gx7KUQbhSqrjQw4AwAO6lb3YHi9EqRGQMRwJXjxEJbGSCOVTZUYVUMADUmu1+oGQw9AFJzaaAISauimgD81oYIAYdFtdrMDuKDHwEFw1RQTegpTAdmpuDW0EddlKsHw2DSjhkmkmdd4qy1ieTAD4gywrNgEAA1bCSDgwUbTu0TOvTewbb0MO7n/wDfyrGzNqAATwAoggVEFeCF3JkNAfpgBfKdj2IREtSqTIF3aX8pkAv85WsWwHBcNxPEQnw3ECYJQgiaJYniRJklSdIOhyOUCiKEoygqHo6gwRpXVaI17HIbpA0zQ1hhNOUjzrB16OWJU1W2XZ9kOY47FOc5RCuG47gwB4nn4GBXneT5QQoKUUHIL4UFWDF/hUMEdKhRwUDoFBwTBOhHHBXFyEROhgG0L5iVJckuSpGlcDpVh72gR9XN5cgZnqDhVlWGRvIAFRgBB5P41Bny4SAJygUUJXIDFZWoeNzEgUJqh6NcEH4Dh8CSlKYGzewNTAnhV3XLcd1KeYUFNK1tFtB1ynIOgkHIUhEW0bQgJnOdtQ4cD2OquwGAAHk1YrSvK5KnCqpiuL/KYeOPMwQK1ViDRm1MGFIU15m0drrWgutYNysworsM9vxuu1mAqYh20e57TXVZdTwgc9LwYa8MFvPyn1fd8sJe+7BzPC8OEkSRXoYWD/yAA). The "throw helper" is used to split the exception logic out, helpful as this method should be inlined.

From https://github.com/dotnet/aspnetcore/pull/31267#discussion_r602692873 (I just can't look at the current code 😉).
In runtimes this is fixed by https://github.com/dotnet/runtime/pull/50549

/cc: @BrennanConroy

### CODE REVIEW

File: src/Http/Headers/src/ContentDispositionHeaderValue.cs [hunk 1]
```diff
@@ -544,16 +544,18 @@ private bool RequiresEncoding(StringSegment input)
...
-            outlen += ((inputLength % 3) != 0) ? 4 : 0;         // at most 4 more chars for the remainder
+            // Based on https://github.com/dotnet/runtime/blob/ac21254184f5a6cec884d8d4049c80d905170b38/src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs#L150
 
-            if (outlen > int.MaxValue)
+            const int MaximumEncodeLength = (int.MaxValue / 4) * 3; // 1610612733
+
+            if ((uint)inputLength <= MaximumEncodeLength)
             {
-                throw new OutOfMemoryException();
+                return ((inputLength + 2) / 3) * 4;
```
Comment (BrennanConroy) [inline src/Http/Headers/src/ContentDispositionHeaderValue.cs:553]: I feel like I'm missing something here. Isn't this the exact same as the method `Base64Encoder.GetMaxEncodedToUtf8Length`? And I thought that method gave you the max, not the exact length.

Comment (gfoidl) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:553]: > Isn't this the exact same as the method 

Yes, except different exception. I'm not sure which exception (OutOfMemory or ArgumentOutOfRange) is the better one -- I'd prefer the latter as used by Base64Encoder.GetMaxEncodedToUtf8Length.

The naming with "max" is IMHO not ideal for _encoding_, as the exact lenght can be computed -- as done by (l + 2) / 3 * 4.
For decoding the exact length can't be computed, due padding, so there "max" makes sense, and I assume the naming was choses for "symmetry".

Would you prefer to just forward to `Base64Encoder.GetMaxEncodedToUtf8Length` or use that method directly?

Comment (Tratcher) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:553]: OutOfMemoryException is pretty strange given an this method doesn't attempt any allocations. It only makes sense if the caller would immediately allocate for it. I'd recomend ArgumentOutOfRange too.

Comment (gfoidl) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs:553]: OK, I'll use `Base64.GetMaxEncodedToUtf8Length` directly, as everything we need is there.

Comment (BrennanConroy) [reply src/Http/Headers/src/ContentDispositionHeaderValue.cs]: Yeah that "Max" is pretty awful in the API and confused me, looks like just using the badly named API is best.

**Review by BrennanConroy — APPROVED** [review]

**Comment by BrennanConroy** [comment]
Thanks @gfoidl

**Pull request merged**